link_ancestors stops after all relevant edges have been processed

gtsambos opened the pull request:
Not finished yet, but shows one way of addressing the efficiency problems mentioned in #2442. Currently assumes that the nodes IDs are in the (canonical) backwards-time order, which is why I changed one of the tests. Otherwise nothing breaks!

Fixes #2442

jeromekelleher commented on python/tests/simplify.py:
```diff
@@ -582,12 +582,15 @@ def __init__(self, ts, sample, ancestors):
             all_edges = list(self.ts.edges())
             edges = all_edges[:1]
             for e in all_edges[1:]:
+                if e.parent > self.oldest_ancestor:
```
I don't think works in general, you're assuming that nodes have nondecreasing time. Better to check ``time[e.parent] > self.oldest_ancestor_time``.

gtsambos replied:
just checking, there isn't a max() function in any of the C libraries tskit relies on, right? if not I'll write that myself

benjeffery replied:
There's python's native `max` or `numpy.ndarray.max`?

jeromekelleher replied:
There's a TSK_MAX macro you can use in a loop.

jeromekelleher reviewed (commented):
Looks good, and quite a simple change to make for a big gain. Better to explicity compare times though.

gtsambos commented:
Thanks Ben! It's the C version of this function that I'm less sure about --
if you could take a look at the function I've added in the latest commit
I'd really appreciate it!

On Thu, 4 Aug 2022, 9:22 pm Ben Jeffery, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In python/tests/simplify.py
> <https://github.com/tskit-dev/tskit/pull/2456#discussion_r937665326>:
>
> >
>      def link_ancestors(self):
>          if self.ts.num_edges > 0:
>              all_edges = list(self.ts.edges())
>              edges = all_edges[:1]
>              for e in all_edges[1:]:
> +                if e.parent > self.oldest_ancestor:
>
> There's python's native max or numpy.ndarray.max?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tskit-dev/tskit/pull/2456#discussion_r937665326>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEHOXQQNJ3ARJMQL6Q2KEK3VXORWTANCNFSM55RJ32GQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

gtsambos commented:
Let me know if this looks ready to go, @benjeffery or @jeromekelleher

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7512,6 +7513,32 @@ ancestor_mapper_add_ancestry(ancestor_mapper_t *self, tsk_id_t input_id, double
+    double max_time = -1;
+
+    for (a = 0; a < self->num_ancestors; a++) {
+        if (node_time[self->ancestors[a]] > max_time) {
```
Could use ``max_time = TSK_MAX(max_time, node_time[self->ancestors[a]]);``

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7512,6 +7513,32 @@ ancestor_mapper_add_ancestry(ancestor_mapper_t *self, tsk_id_t input_id, double
+            max_time = node_time[self->samples[a]];
+        }
+    }
+    if (max_time < 0) {
```
I guess this can only happen if we have zero samples and zero ancestors, which we're already checking for?

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7818,6 +7850,10 @@ ancestor_mapper_run(ancestor_mapper_t *self)
                 }
                 current_parent = parent;
+                if (node_time[current_parent] > self->oldest_node_time) {
+                    goto out;
```
We only use ``goto out`` for error conditions. You can ``break`` here instead, right?

gtsambos replied:
I don't think we can use break, because we have more than one loop to break out of. We end up getting other errors:
```
stderr:
Bug detected in ../tskit/tables.c at line 7802. If you are using tskit directly please open an issue on GitHub, ideally with a reproducible example. (https://github.com/tskit-dev/tskit/issues) If you are using software that uses tskit, please report an issue to that software's issue tracker, at least initially.
```

gtsambos replied:
ok -- I put that final call inside another if statement

jeromekelleher commented on python/tests/simplify.py:
```diff
@@ -582,12 +582,17 @@ def __init__(self, ts, sample, ancestors):
         for sample_id in sample:
             self.add_ancestry(0, self.sequence_length, sample_id, sample_id)
         self.edge_buffer = {}
+        self.oldest_ancestor_time = max(ts.tables.nodes.time[u] for u in ancestors)
```
``ts.nodes_time[u]`` is the preferred approach now - it's a lot faster. Ditto below.

jeromekelleher reviewed (commented):
LGTM, a few minor implementation points.

jeromekelleher commented:
Can you update the CHANGELOG also please @gtsambos?

jeromekelleher commented:
The docs fail is probably transient, don't worry about it.

benjeffery commented:
I think the docs might need a fix, I'll get on it.

gtsambos commented:
the codecov gods are cruel and unforgiving! ⚡

jeromekelleher commented on c/tskit/tables.c:
```diff
@@ -7818,13 +7838,19 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+                    break;
+                }
+
                 start = j;
```
I think we can make this a bit simpler - why do we have the last call to ancestor_mapper_process_parent_edges below? 

I think you should put the ``break`` after the ``start = j`` bit also, to keep the loop invariants correct.

gtsambos replied:
I guess it depends on what makes things simpler -- the latest push changes it so that process_parent_edges is only called inside the loop, but it means everything in the loop has to be put inside conditional statements.

gtsambos commented on c/tskit/tables.c:
```diff
@@ -7805,27 +7824,29 @@ ancestor_mapper_run(ancestor_mapper_t *self)
-        for (j = 0; j < num_edges; j++) {
-            parent = input_edges->parent[j];
-            if (parent != current_parent) {
+        parent = input_edges->parent[0];
```
Note this line isn't strictly necessary, but without it gcc now throws a warning about using `parent` uninitialised (probably because the initialisation is happening inside an if statement now)

jeromekelleher commented:
I've added  a suggested refactoring here that avoids within-loop checks (which I find a bit confusing).

You seem to have made changes to the kastore submodule here which will need to be dropped.

gtsambos commented:
hmm, Jerome's commit credit was removed when I squashed and rebased the commit, somehow 😮‍💨 you can see from the commit message that his code is in there, though

jeromekelleher commented on python/tests/simplify.py:
```diff
@@ -1,6 +1,6 @@
 # MIT License
 #
-# Copyright (c) 2018-2021 Tskit Developers
+# Copyright (c) 2022 Tskit Developers
```
Something went wrong here - should be 2019-2022

jeromekelleher reviewed (approved):
Looks great, thanks @gtsambos! 

One tiny nit on the copyright updates, but happy to merge whenever.

jeromekelleher commented:
> hmm, Jerome's commit credit was removed when I squashed and rebased the commit, somehow ￼ you can see from the commit message that his code is in there, though

That's fine, as I expected.
